internal/rows/arrowbased/arrowRows.go

```diff
@@ -0,0 +1,1158 @@
+		col.Retain()
+		defer col.Release()
+
+		colData := col.Data()
+		colData.Retain()
+		defer col.Release()
+
+		colValsHolder := ars.columnValues[i]
+
+		// release the arrow array already held
+		colValsHolder.Release()
+
+		err := colValsHolder.SetValueArray(colData)
+		if err != nil {
+			ars.Error().Msg(err.Error())
+		}
+	}
+
+	ars.currentBatch = &ars.arrowBatches[batchIndex]
+
+	return nil
+}
+
+// getArrowSchemaBytes returns the serialized schema in ipc format
+func getArrowSchemaBytes(schema *arrow.Schema) ([]byte, error) {
+	if schema == nil {
+		return nil, errors.New(errArrowRowsNilArrowSchema)
+	}
+
+	var output bytes.Buffer
+	w := ipc.NewWriter(&output, ipc.WithSchema(schema))
+	err := w.Close()
+	if err != nil {
+		return nil, dbsqlerr.WrapErr(err, errArrowRowsUnableToWriteArrowSchema)
+	}
+
+	arrowSchemaBytes := output.Bytes()
+
+	// the writer serializes to an arrow batch but we just want the
+	// schema bytes so we strip off the empty Record at the end
+	arrowSchemaBytes = arrowSchemaBytes[:len(arrowSchemaBytes)-8]
+
+	return arrowSchemaBytes, nil
+}
+
+// rowIndexToBatchIndex returns the index of the batch containing the specified row
+func (ars *arrowRowScanner) rowIndexToBatchIndex(rowIndex int64) (int, error) {
+
+	for i := range ars.arrowBatches {
+		if ars.arrowBatches[i].contains(rowIndex) {
+			return i, nil
+		}
+	}
+
+	ars.Error().Msgf(errArrowRowsInvalidRowIndex, rowIndex)
+	return -1, errors.Errorf(errArrowRowsInvalidRowIndex, rowIndex)
+}
+
+// tTableSchemaToArrowSchema convers the TTableSchema retrieved by the thrift server into an arrow.Schema instance
+func tTableSchemaToArrowSchema(schema *cli_service.TTableSchema, arrowConfig *config.ArrowConfig) (*arrow.Schema, error) {
+	columns := schema.GetColumns()
+	fields := make([]arrow.Field, len(columns))
+
+	for i := range columns {
+		field, err := tColumnDescToArrowField(columns[i], arrowConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		fields[i] = field
+	}
+
+	arrowSchema := arrow.NewSchema(fields, nil)
+
+	return arrowSchema, nil
+}
+
+// map the thrift data types to the corresponding arrow data type
+var toArrowTypeMap map[cli_service.TTypeId]arrow.DataType = map[cli_service.TTypeId]arrow.DataType{
+	cli_service.TTypeId_BOOLEAN_TYPE:  arrow.FixedWidthTypes.Boolean,
+	cli_service.TTypeId_TINYINT_TYPE:  arrow.PrimitiveTypes.Int8,
+	cli_service.TTypeId_SMALLINT_TYPE: arrow.PrimitiveTypes.Int16,
+	cli_service.TTypeId_INT_TYPE:      arrow.PrimitiveTypes.Int32,
+	cli_service.TTypeId_BIGINT_TYPE:   arrow.PrimitiveTypes.Int64,
+	cli_service.TTypeId_FLOAT_TYPE:    arrow.PrimitiveTypes.Float32,
+	cli_service.TTypeId_DOUBLE_TYPE:   arrow.PrimitiveTypes.Float64,
+	cli_service.TTypeId_STRING_TYPE:   arrow.BinaryTypes.String,
+	// cli_service.TTypeId_TIMESTAMP_TYPE:    see tColumnDescToArrowDataType
+	cli_service.TTypeId_BINARY_TYPE: arrow.BinaryTypes.Binary,
+	// cli_service.TTypeId_ARRAY_TYPE:        see tColumnDescToArrowDataType
```

Comment:
The thought was that someone looking at the map would realize that not all thrift types have an entry, because they don't map to a simple arrow data type, and to direct them to where those types are handled